### PR TITLE
TCP reassembly buffer v184.4

### PR DESCRIPTION
**Pull request by inliniac**
Switch over TCP engine to use streaming buffer API. Major rewrite and cleanup.

- Gets rid of the StreamMsg's and instead inspects the streaming buffer.
- Improve overlap detection accuracy
- simplify config: no more per size segment pools

### CODE REVIEW

**Comment by pevma** [comment]
What is the size of each prealloc segment (from segment-prealloc)?

**Comment by inliniac** [comment]
40 bytes (on 64bit). This does not include the segment data/payload, which is now stored in the stream directly.

File: src/flow-manager.c [hunk 1]
```diff
@@ -1200,7 +1200,7 @@ static int FlowMgrTest02 (void)
     struct timeval ts;
     TcpSegment seg;
     TcpStream client;
-    uint8_t payload[3] = {0x41, 0x41, 0x41};
+    //uint8_t payload[3] = {0x41, 0x41, 0x41};
```
Comment (inliniac) [inline src/flow-manager.c]: clean up

File: src/stream-tcp-list.c [hunk 2]
```diff
@@ -0,0 +1,954 @@
...
+
+        /* now let's consider the data in the overlap case */
+        int res = DoHandleData(tv, ra_ctx, stream, seg, p);
+        if (res < 0) {
+            StatsIncr(tv, ra_ctx->counter_tcp_reass_data_overlap_fail);
+            StreamTcpRemoveSegmentFromStream(stream, seg);
+            StreamTcpSegmentReturntoPool(seg);
+            SCReturnInt(-1);
+        }
+
```
Comment (inliniac) [inline src/stream-tcp-list.c]: minor: style

File: src/stream-tcp-private.h [hunk 3]
```diff
@@ -51,16 +52,21 @@ typedef struct StreamTcpSackRecord_ {
...
+    /* coccinelle: TcpSegment:flags:SEGMENTTCP_FLAG */
+    uint8_t flags;
     uint32_t seq;
     struct TcpSegment_ *next;
     struct TcpSegment_ *prev;
-    /* coccinelle: TcpSegment:flags:SEGMENTTCP_FLAG */
-    uint8_t flags;
 } TcpSegment;
 
+#define TCP_SEG_LEN(seg)        (seg)->payload_len
```
Comment (inliniac) [inline src/stream-tcp-private.h]: reconsider

File: src/stream-tcp-reassemble.c [hunk 4]
```diff
@@ -300,286 +336,74 @@ int StreamTcpAppLayerIsDisabled(Flow *f)
...
+        StreamTcpReassembleConfigEnableOverlapCheck();
+    }
+    if (StreamTcpInlineMode() == TRUE) {
+        StreamTcpReassembleConfigEnableOverlapCheck();
     }
-    stream_config.zero_copy_size = (uint16_t)zero_copy_size;
-    if (!quiet)
-        SCLogConfig("stream.reassembly \"zero-copy-size\": %u", stream_config.zero_copy_size);
+
+    /** \todo do this for real */
```
Comment (inliniac) [inline src/stream-tcp-reassemble.c]: remove

**Comment by inliniac** [comment]
Replaced by #2656

**Pull request closed without merging**